src/lib/drivers/accelerometer/PX4Accelerometer.cpp

```diff
@@ -0,0 +1,150 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2018 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+
+#include "PX4Accelerometer.hpp"
+
+#include <lib/drivers/device/Device.hpp>
+
+PX4Accelerometer::PX4Accelerometer(uint32_t device_id, uint8_t priority, enum Rotation rotation) :
+	CDev(nullptr),
+	ModuleParams(nullptr),
+	_sensor_accel_pub{ORB_ID(sensor_accel), priority},
+	_rotation{get_rot_matrix(rotation)}
+{
+	_class_device_instance = register_class_devname(ACCEL_BASE_DEVICE_PATH);
+
+	_sensor_accel_pub.get().device_id = device_id;
+	_sensor_accel_pub.get().scaling = 1.0f;
+
+	// set software low pass filter for controllers
+	updateParams();
+	configure_filter(_filter_cutoff.get());
+
+	// force initial publish to allocate uORB buffer
+	// TODO: can be removed once all drivers are in threads
+	_sensor_accel_pub.update();
```

Comment:
The initial publish (advertise) could fail if out of memory, but even now in master that's only handled with a warning in init. At least when using the uORB publication wrappers like this it will try the advertise again next publish.

Reply:
Since this will be a repeated pattern and constructors can not return a value. I would separate the construtor and the things that can return errors. This gives us a LCD for the embedded no exception model.